### PR TITLE
Make PyDictionary and PyList convertable back to PyObject without marshalling

### LINKED ISSUE #199: Console sample invalid cast exception
On a fresh local clone on [main here] (https://github.com/tonybaloney/CSnakes/commit/949b73839fe938a678859dc999e7a3cb92c1e3ce), I use VS to build and run in debug mode the Quick Console Test solution and get this error:

```bash
System.InvalidCastException
  HResult=0x80004002
  Message=Cannot convert [a, 1] to PyObject
  Source=CSnakes.Runtime
  StackTrace:
   at CSnakes.Runtime.Python.PyObject.From[T](T value)
   at CSnakes.Runtime.PyObjectTypeConverter.ConvertFromList(IEnumerable e)
   at CSnakes.Runtime.Python.PyObject.From[T](T value)
   at CSnakes.Runtime.TypeDemosExtensions.TypeDemosInternal.TakeDict(IReadOnlyDictionary`2 d) in C:\Users\sarahkaiser\git\CSnakes\samples\simple\ExamplePythonDependency\obj\Debug\net8.0\generated\CSnakes\CSnakes.PythonStaticGenerator\TypeDemos.py.cs:line 72
   at Program.<<Main>$>g__RunDictionaryDemo|0_2(IPythonEnvironment env) in C:\Users\sarahkaiser\git\CSnakes\samples\simple\QuickConsoleTest\Program.cs:line 51
   at Program.<Main>$(String[] args) in C:\Users\sarahkaiser\git\CSnakes\samples\simple\QuickConsoleTest\Program.cs:line 30
```

Console output:

```bash
info: CSnakes.Runtime.IPythonEnvironment[0]
      Setting up Python environment from C:\Users\sarahkaiser\.nuget\packages\python\3.12.4\tools using home of C:\Users\sarahkaiser\git\CSnakes\samples\simple\QuickConsoleTest\bin\Debug\net8.0\..\..\..\..\ExamplePythonDependency
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      File C:\Users\sarahkaiser\git\CSnakes\samples\simple\QuickConsoleTest\bin\Debug\net8.0\..\..\..\..\ExamplePythonDependency\requirements.txt was found.
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Using virtual environment at C:\Users\sarahkaiser\git\CSnakes\samples\simple\QuickConsoleTest\bin\Debug\net8.0\..\..\..\..\ExamplePythonDependency\.venv to install packages with pip.
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: numpy in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from -r requirements.txt (line 1)) (2.1.1)
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: scikit-learn in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from -r requirements.txt (line 2)) (1.5.2)
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: wheel in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from -r requirements.txt (line 3)) (0.44.0)
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: transformers in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from -r requirements.txt (line 4)) (4.44.2)
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: scipy>=1.6.0 in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from scikit-learn->-r requirements.txt (line 2)) (1.14.1)
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: joblib>=1.2.0 in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from scikit-learn->-r requirements.txt (line 2)) (1.4.2)
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: threadpoolctl>=3.1.0 in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from scikit-learn->-r requirements.txt (line 2)) (3.5.0)
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: filelock in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from transformers->-r requirements.txt (line 4)) (3.16.0)
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: huggingface-hub<1.0,>=0.23.2 in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from transformers->-r requirements.txt (line 4)) (0.24.7)
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: packaging>=20.0 in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from transformers->-r requirements.txt (line 4)) (24.1)
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: pyyaml>=5.1 in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from transformers->-r requirements.txt (line 4)) (6.0.2)
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: regex!=2019.12.17 in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from transformers->-r requirements.txt (line 4)) (2024.9.11)
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: requests in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from transformers->-r requirements.txt (line 4)) (2.32.3)
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: safetensors>=0.4.1 in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from transformers->-r requirements.txt (line 4)) (0.4.5)
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: tokenizers<0.20,>=0.19 in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from transformers->-r requirements.txt (line 4)) (0.19.1)
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: tqdm>=4.27 in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from transformers->-r requirements.txt (line 4)) (4.66.5)
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: fsspec>=2023.5.0 in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from huggingface-hub<1.0,>=0.23.2->transformers->-r requirements.txt (line 4)) (2024.9.0)
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: typing-extensions>=3.7.4.3 in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from huggingface-hub<1.0,>=0.23.2->transformers->-r requirements.txt (line 4)) (4.12.2)
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: colorama in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from tqdm>=4.27->transformers->-r requirements.txt (line 4)) (0.4.6)
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: charset-normalizer<4,>=2 in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from requests->transformers->-r requirements.txt (line 4)) (3.3.2)
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: idna<4,>=2.5 in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from requests->transformers->-r requirements.txt (line 4)) (3.10)
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: urllib3<3,>=1.21.1 in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from requests->transformers->-r requirements.txt (line 4)) (2.2.3)
info: CSnakes.Runtime.PackageManagement.PipInstaller[0]
      Requirement already satisfied: certifi>=2017.4.17 in c:\users\sarahkaiser\git\csnakes\samples\simple\examplepythondependency\.venv\lib\site-packages (from requests->transformers->-r requirements.txt (line 4)) (2024.8.30)
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaa,bbb,ccc
a : 1
b : 2
c : 3
Unhandled exception. System.InvalidCastException: Cannot convert [a, 1] to PyObject
   at CSnakes.Runtime.Python.PyObject.From[T](T value) in /_/src/CSnakes.Runtime/Python/PyObject.cs:line 509
   at CSnakes.Runtime.PyObjectTypeConverter.ConvertFromList(IEnumerable e) in /_/src/CSnakes.Runtime/PyObjectTypeConverter.List.cs:line 40
   at CSnakes.Runtime.Python.PyObject.From[T](T value) in /_/src/CSnakes.Runtime/Python/PyObject.cs:line 507
   at CSnakes.Runtime.TypeDemosExtensions.TypeDemosInternal.TakeDict(IReadOnlyDictionary`2 d) in C:\Users\sarahkaiser\git\CSnakes\samples\simple\ExamplePythonDependency\obj\Debug\net8.0\generated\CSnakes\CSnakes.PythonStaticGenerator\TypeDemos.py.cs:line 72
   at Program.<<Main>$>g__RunDictionaryDemo|0_2(IPythonEnvironment env) in C:\Users\sarahkaiser\git\CSnakes\samples\simple\QuickConsoleTest\Program.cs:line 51
   at Program.<Main>$(String[] args) in C:\Users\sarahkaiser\git\CSnakes\samples\simple\QuickConsoleTest\Program.cs:line 30
```
It should be using the Python env that the C# toolchain installs, not local to my machine, but digging in now.

### PR DESCRIPTION
**Pull request by tonybaloney**
Fixes #199

### CODE REVIEW

File: src/CSnakes.Runtime/Python/ICloneable.cs
Comment (aaronpowell) [inline src/CSnakes.Runtime/Python/ICloneable.cs:1]: Just use https://learn.microsoft.com/en-us/dotnet/api/system.icloneable?view=net-8.0

File: src/CSnakes.Runtime/Python/PyDictionary.cs [hunk 1]
```diff
@@ -100,5 +100,10 @@ public bool TryGetValue(TKey key, [MaybeNullWhen(false)] out TValue value)
         return false;
     }
 
+    PyObject ICloneable.Clone()
+    {
+        return _dictionaryObject.Clone();
+    }
+
```
Comment (aaronpowell) [inline src/CSnakes.Runtime/Python/PyDictionary.cs:107]:
```suggestion
    PyObject ICloneable.Clone() => dictionaryObject.Clone();
```

File: src/CSnakes.Runtime/Python/PyList.cs [hunk 2]
```diff
@@ -37,21 +33,26 @@ public int Count
...
-            return new PyEnumerable<TItem>(_listObject);
+            return new PyEnumerable<TItem>(listObject);
         }
     }
 
+    PyObject ICloneable.Clone()
+    {
+        return listObject.Clone();
+    }
+
```
Comment (aaronpowell) [inline src/CSnakes.Runtime/Python/PyList.cs:56]:
```suggestion
    PyObject ICloneable.Clone() => listObject.Clone();
```

File: src/CSnakes.Runtime/Python/PyObject.cs [hunk 3]
```diff
@@ -541,4 +541,9 @@ private static void MergeKeywordArguments(string[] kwnames, PyObject[] kwvalues,
         combinedKwnames = [.. newKwnames];
         combinedKwvalues = [.. newKwvalues];
     }
+
+    PyObject ICloneable.Clone()
+    {
+        return Clone();
+    }
```
Comment (aaronpowell) [inline src/CSnakes.Runtime/Python/PyObject.cs:548]:
```suggestion
    PyObject ICloneable.Clone() => Clone();
```

File: src/Integration.Tests/DictsTests.cs [hunk 4]
```diff
@@ -43,4 +43,17 @@ public void TestDicts_TestMapping()
         Assert.Equal(1, result["dictkey1"]);
         Assert.Equal(2, result["dictkey2"]);
     }
+
+    [Fact]
+    public void TestDictionaryRoundTrip()
```
Comment (aaronpowell) [inline src/Integration.Tests/DictsTests.cs:48]: Should we have a test that the objects aren't the same?

**Review by aaronpowell — CHANGES_REQUESTED** [review]

File: src/CSnakes.Runtime/Python/ICloneable.cs
Comment (tonybaloney) [reply src/CSnakes.Runtime/Python/ICloneable.cs:1]: I don't want to match _anything_ that can be cloned. I want to match anything that can be cloned to `PyObject`

**Pull request merged**